### PR TITLE
Implement `search_traces`

### PR DESCRIPTION
**Pull request by harupy**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/harupy/mlflow/pull/11578?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11578/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11578
```

</p>
</details>

### Related Issues/PRs

 #xxx

### What changes are proposed in this pull request?

Title

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

```python
import time
from unittest import mock

import requests

import mlflow

mlflow.set_tracking_uri("databricks")
name = "/Users/harutaka.kawamura@databricks.com/test"
mlflow.set_experiment(name)
exp = mlflow.get_experiment_by_name(name)

print(exp)

client = mlflow.MlflowClient()
trace_info = client.create_trace(
    experiment_id=exp.experiment_id,
    timestamp_ms=123,
    execution_time_ms=567,
    status="OK",
)

print(trace_info)
res = client.search_traces(experiment_ids=[exp.experiment_id], max_results=10)
print(res)
res = client.search_traces(experiment_ids=[exp.experiment_id], max_results=10, page_token=res.token)
print(res)

ts = int(time.time() * 1000)
trace_info = client.create_trace(
    experiment_id=exp.experiment_id,
    timestamp_ms=ts + 10,
    execution_time_ms=567,
    status="OK",
)
res = client.search_traces(
    experiment_ids=[exp.experiment_id],
    filter_string=f"trace.timestamp_ms > {ts}",
)
print(res)

trace_info = client.create_trace(
    experiment_id=exp.experiment_id,
    timestamp_ms=123,
    execution_time_ms=567,
    status="ERROR",
)

o = requests.Session.request


def request(*args, **kwargs):
    res = o(*args, **kwargs)
    print(res.json())
    return res


with mock.patch("requests.Session.request", request):
    res = client.search_traces(
        experiment_ids=[exp.experiment_id],
        filter_string="trace.status = 'ERROR'",
    )
    print(res)
```

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/tracking/client.py [hunk 1]
```diff
@@ -444,6 +444,41 @@ def get_trace_info(self, request_id: str) -> TraceInfo:
...
         return self._tracking_client.get_trace_info(request_id)
 
+    def search_traces(
+        self,
+        experiment_ids: List[str],
+        filter_string: Optional[str] = None,
+        max_results: int = SEARCH_TRACES_MAX_RESULTS,
+        order_by: Optional[List[str]] = None,
+        page_token: Optional[str] = None,
+    ):
```
Comment (dbczumar) [inline mlflow/tracking/client.py:454]: The client API should return `Trace`s  including `TraceData`, not just `TraceInfo`. For that reason, we should probably start with a default max results of ~ 100 traces.

Comment (dbczumar) [reply mlflow/tracking/client.py:454]: Yeah! :)

File: mlflow/tracking/_tracking_service/client.py [hunk 2]
```diff
@@ -198,6 +204,30 @@ def get_trace_info(self, request_id):
...
+            filter_string=filter_string,
+            max_results=max_results,
+            order_by=order_by,
+            page_token=page_token,
+        )
+        traces = [
+            Trace(
+                trace_info=trace_info,
+                trace_data=self._download_trace_data(trace_info.request_id),
+            )
```
Comment (harupy) [inline mlflow/tracking/_tracking_service/client.py:226]: This can be parallelized.

File: mlflow/store/tracking/__init__.py [hunk 3]
```diff
@@ -19,3 +19,4 @@
 SEARCH_MAX_RESULTS_DEFAULT = 1000
 SEARCH_MAX_RESULTS_THRESHOLD = 50000
 GET_METRIC_HISTORY_MAX_RESULTS = 25000
+SEARCH_TRACES_MAX_RESULTS = 500
```
Comment (harupy) [inline mlflow/store/tracking/__init__.py:22]:
```suggestion
SEARCH_TRACES_MAX_RESULTS = 100
```

File: mlflow/tracking/_tracking_service/client.py (hunk 2 above)
Comment (dbczumar) [reply mlflow/tracking/_tracking_service/client.py:226]: Can we implement that?

**Review by dbczumar — APPROVED** [review]
LGTM once we've implemented parallelism for downloading trace data

File: mlflow/tracking/client.py [hunk 4]
```diff
@@ -444,6 +444,41 @@ def get_trace_info(self, request_id: str) -> TraceInfo:
...
         return self._tracking_client.get_trace_info(request_id)
 
+    def search_traces(
+        self,
+        experiment_ids: List[str],
+        filter_string: Optional[str] = None,
+        max_results: int = SEARCH_TRACES_DEFAULT_MAX_RESULTS,
+        order_by: Optional[List[str]] = None,
+        page_token: Optional[str] = None,
+    ):
```
Comment (dbczumar) [inline mlflow/tracking/client.py:454]: @harupy One more request: Can we add a fluent API?

File: mlflow/tracking/_tracking_service/client.py (hunk 2 above)
Comment (harupy) [reply mlflow/tracking/_tracking_service/client.py:226]: can we implement this as a follow-up?

Comment (dbczumar) [reply mlflow/tracking/_tracking_service/client.py:226]: What's the performance look like currently? If it's really slow, we should do it now

File: mlflow/tracking/client.py (hunk 4 above)
Comment (dbczumar) [reply mlflow/tracking/client.py:456]: ^ It's fine to do this as a follow-up

File: mlflow/tracking/_tracking_service/client.py (hunk 2 above)
Comment (harupy) [reply mlflow/tracking/_tracking_service/client.py]:
```
res = client.search_traces(experiment_ids=[exp.experiment_id], max_results=50)
```

took 45 seconds on my machine. Implementing parallel downloads now.

**Pull request merged**